app/lib/advanced_text_formatter.rb

```diff
@@ -27,73 +29,95 @@ def autolink(link, link_type)
+      when 'hashtag'
+        entity.replace(link_to_hashtag({ hashtag: entity['value'] }))
+      when 'link'
+        entity.replace(link_to_url({ url: entity['value'] }))
+      when 'mention'
+        entity.replace(link_to_mention({ screen_name: entity['value'] }))
       end
     end
-
-    html.html_safe # rubocop:disable Rails/OutputSafety
+    result.to_html.html_safe # rubocop:disable Rails/OutputSafety
   end
 
-  # Differs from `TextFormatter` by skipping HTML tags and entities
-  def entities
-    @entities ||= begin
-      gaps = []
-      total_offset = 0
-
-      escaped = text.gsub(/<[^>]*>|&#[0-9]+;/) do |match|
-        total_offset += match.length - 1
-        end_offset = Regexp.last_match.end(0)
-        gaps << [end_offset - total_offset, total_offset]
-        ' '
-      end
-
-      Extractor.extract_entities_with_indices(escaped, extract_url_without_protocol: false).map do |entity|
-        start_pos, end_pos = entity[:indices]
-        offset_idx = gaps.rindex { |gap| gap.first <= start_pos }
-        offset = offset_idx.nil? ? 0 : gaps[offset_idx].last
-        entity.merge(indices: [start_pos + offset, end_pos + offset])
+  ##
+  # Process the status into a Nokogiri document fragment, with entities
+  # replaced with +<mastodon-entity>+s.
+  #
+  # Since +<mastodon-entity>+ is not allowed by the sanitizer, any such
+  # elements in the output *must* have been produced by this algorithm.
+  #
+  # These elements will need to be replaced prior to serialization (see
+  # +#to_s+).
+  def tree
+    if @tree.nil?
+      src = text.gsub(Sanitize::REGEX_UNSUITABLE_CHARS, '')
+      @tree = Nokogiri::HTML5.fragment(src)
+      Sanitize.node!(@tree, Sanitize::Config::MASTODON_OUTGOING)
+      document = @tree.document
+
+      @tree.xpath('.//text()[not(ancestor::a | ancestor::code)]').each do |text_node|
+        # Iterate over text elements and build up their replacements.
+        content = text_node.content
+        replacement = Nokogiri::XML::NodeSet.new(document)
+        processed_index = 0
+        Extractor.extract_entities_with_indices(
+          content,
+          extract_url_without_protocol: false
+        ) do |entity|
+          # Iterate over entities in this text node.
+          advance = entity[:indices].first - processed_index
+          if advance.positive?
+            # Text node for content which precedes entity.
+            replacement << Nokogiri::XML::Text.new(
+              content[processed_index, advance],
+              document
+            )
+          end
+          elt = Nokogiri::XML::Element.new('mastodon-entity', document)
+          if entity[:url]
+            elt['kind'] = 'link'
+            elt['value'] = entity[:url]
+          elsif entity[:hashtag]
+            elt['kind'] = 'hashtag'
+            elt['value'] = entity[:hashtag]
+          elsif entity[:screen_name]
+            elt['kind'] = 'mention'
+            elt['value'] = entity[:screen_name]
```

Comment:
Is there a reason not to output the replaced entity directly?

Reply:
this was a conservative approach based on me not being entirely sure how this formatter was used. i noticed that the basic text formatter computes `@entities` once, and then redoes the entity substitutions on every call to `#to_s`. i tried to follow this pattern by computing `@tree` once but then likewise performing the substitutions on every call to `#to_s`.

it could be that that’s entirely unnecessary and the substitutions could just be done when setting `@tree`; i just wasn’t sure :P . let me know what you think!

Reply:
Yeah, I'd prefer doing it when setting `@tree`, for performance issues, and in order to run the sanitizer on the final markup instead of on an intermediary one (even if the nodes we replace after are considered safe).

Reply:
Text nodes and replaced entities should be left unchanged by the sanitizer, so I don't think there's a strong reason to run it before the entity replacement code. So I think for performance reasons it's desirable to run it only once, and for security reasons this should happen at the end.

Reply:
I took the liberty of pushing these changes to the PR. I'll try it out for a while, and possibly add more tests, but if nothing comes up this looks good to me! You're of course welcome to review that change in the meantime!

Reply:
(sorry for the slow responses; U.S. holidays)

The case I was considering when sanitizing at the beginning was like:

```html
<p>Some text with a @men<unsupported-element>tion</unsupported-element></p>
```

If you sanitize first, then that becomes `@mention`—but you’re probably right that it is still broken across two text nodes, so maybe behaviour isn’t actually different after all. Considering that, I’m happy with these changes.